Success sign up with a github account using oauth

jangjichang opened the pull request:
Created a function based view register as a github account by referring
to 'https://developer.github.com/apps/building-oauth-apps/
authorizing-oauth-apps/'.

I'm going to refactor with 'TODO' written in the github_login_callback
function.

jangjichang committed:
Success sign up with a github account using oauth

Created a function based view register as a github account by referring
to 'https://developer.github.com/apps/building-oauth-apps/
authorizing-oauth-apps/'.

I'm going to refactor with 'TODO' written in the github_login_callback
function.

geusan commented on account/views.py:
```diff
@@ -1,11 +1,93 @@
+    return Response({'token': token.key})
+
+
+def making_url(url, params):
```
urllib.parse.urlencode 이걸 이용하면 쿼리스트링 만들어주는데, 내장 라이브러리를 써보면 어떨까? 의견 내봐
[https://docs.python.org/3/library/urllib.parse.html](https://docs.python.org/3/library/urllib.parse.html)

jangjichang replied:
좋습니다! urlencode를 알았다면 making_url 함수를 만들지 않았을 것 같네요. 그렇게 바꿔야겠네요. 👍

geusan commented on account/views.py:
```diff
@@ -1,11 +1,93 @@
+        else:
+            return Response({'message' :'This email has already been used to login ' + user.login_method + ' Please login using ' + user.login_method}, status=status.HTTP_400_BAD_REQUEST)
+    except MyUser.DoesNotExist:
+        avatar = requests.get(user_json['avatar_url'])
```
avatar 이미지는 굳이 파일로 저장할 필요 있는 거야?
깃헙 내에서 수정이 가능하기 때문에, 파일 저장보다는 avatar_url 그대로 저장하는게 관리 리소스를 절약할 수 있지 않을까?

jangjichang replied:
AbstractUser를 상속받아 MyUser 클래스를 만들었어요.
- https://github.com/prography/5th-cogether-back/blob/master/account/models.py#L43

회원가입 방식이 크게 두가지인데요,

첫번째는 위의 소스코드에 29~33번째 줄에 작성한 것 처럼 naver, github, google을 이용한 소셜 로그인이 있고
두번째는 소셜 로그인을 이용하지 않고 이메일과 비밀번호로 회원가입을 할 수도 있어요.

파일로 저장한 이유는 두번째 방식처럼 소셜 로그인을 이용하지 않은 경우에 avatar 이미지를 설정하기 위해 파일로 저장했어요

```python
avatar = models.ImageField(upload_to=uuid_name_upload_to, null=True, blank=True)
```
현재 위와 같은 코드를 아래와 같이 바꾸면 문제가 해결되지 않을까요? 소셜 로그인으로 가입한 경우 avatar 필드에 값을 넣고 이메일로 가입한 경우 profile_picture 필드에 값을 넣도록하면 될 것 같아요.

```python
avatar = models.URLField(blank=True)
profile_picture = models.ImageField(blank=True, null=True)
```

Eunhyang replied:
근데 프로필 이미지가 필요한가 싶기도하넹 :)
깃헙 로그인 안했을 경우에도 s3에 저장해서 url만 저장해도 될것 같기도하고 (돈이 많이 드려나?)

geusan replied:
내 말의 요지는 왜 파일로 저장하냐였어. 서비스  제공업체에서 프로필 이미지에 대한 정보를 url로 주는데, 이걸 그대로 쓰는게 좋지 않을까해서

Eunhyang replied:
응응 그니까 깃헙 이미지도 url로 깃헙로그인 안한 사용자도 s3에 올리고 참조 url로 해서 url필드만 있어도 되지않을까 하는 제안..

geusan replied:
근데 직접 로그인했을 경우에 프로필 이미지를 업로드 해서 필드가 이미지 였구나! 그러면 굳이 안바꿔도 될거같아. 지금 그대로 하고, uuid를 매번 새로 발급하기 보다 유저정보 기반으로 생성하면, 같은 이미지가 중복저장되는 이슈랑, 수정했을때 이전 이미지 삭제해야하는 이슈를 해결 할 수 있을 거 같아!

geusan replied:
나 전에 했을때는, 그냥 프론트에서 알아서 업로드하고 url로만 줘서 고민해본적 없는 거같아ㅋㅋ s3 말고도 꽁짜로 해주는 서비스가 있더라구

jangjichang replied:
소셜 로그인용 아바타 필드랑 이메일 로그인용 아바타 필드를 만들었어요.

프로필 사진 저장 경로도 media/account/myuser/<userid>/userid.jpg 이런식으로 저장되도록 변경했어요~

geusan commented:
밤에 눈팅하다 갑니다 총총

jangjichang committed:
Delete 'making_url' function

Replace the 'making_url' function with the standard library.

jangjichang committed:
Merge branch 'master' into social-login

Eunhyang reviewed (approved):
짝짝짝 멋지구려 👍

jangjichang committed:
Add social_avatar field on MyUser model

A social_avatar field is needed to save social avatar url.

A avatar field is needed to save profile picture for anyone who signs up
by email.

jangjichang committed:
Merge branch 'social-login' of https://github.com/prography/5th-cogether-back into social-login
